src/Parser/TypeParser.php

```diff
@@ -208,4 +219,63 @@ private function tryParseArray(TokenIterator $tokens, Ast\Type\TypeNode $type):
 		return $type;
 	}
 
+
+	private function parseArrayShape(TokenIterator $tokens, Ast\Type\TypeNode $type): Ast\Type\TypeNode
+	{
+		$tokens->consumeTokenType(Lexer::TOKEN_OPEN_CURLY_BRACKET);
+		$items = [$this->parseArrayShapeItem($tokens)];
+
+		while (!$tokens->tryConsumeTokenType(Lexer::TOKEN_CLOSE_CURLY_BRACKET)) {
```

Comment:
for consistency with `parseGeneric` a (micro) speed improvements, this should be written as

```php
while ($tokens->tryConsumeTokenType(Lexer::TOKEN_COMMA)) {
  $items[] = $this->parseArrayShapeItem($tokens);
}

$tokens->consumeTokenType(Lexer::TOKEN_CLOSE_CURLY_BRACKET);
```